File: java/src/Test5.java
import java.util.Scanner;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.io.*;
import java.util.NoSuchElementException;

public class Test5 extends Test {
    /**
     * runTest: run the test specified in the file
     *
     * Arguments:
     *   f: a file containing the test
     *   version: true, if more than pass/fail should be printed
     *
     * Returns: True, if the test passed, failed otherwise.
     **/
    public boolean runTest(File f, boolean verbose) {
	String inputFilename = f.getPath();
	int line = 0;
	HashMap<String, Restaurant> restaurants = new HashMap<String, Restaurant>();
	int unconfirmed = 0;
	int numRestaurants = 0;
	int numReservations = 0;
	int i;
	Restaurant r;

	try {
	    Scanner scanner = new Scanner(f);
	    String[] parameters = scanner.nextLine().split(" ");
	    numRestaurants = Integer.parseInt(parameters[0]);
	    numReservations = Integer.parseInt(parameters[1]);
	    String[] data;

	    for (i = 0; i < numRestaurants; i++) {
		line++;
		data = scanner.nextLine().split(" ");
		r = new Restaurant(data[0], Integer.parseInt(data[1]));
		restaurants.put(data[0], r);
	    }

	    for (i = 0; i < numReservations; i++) {
		line++;		
		data = scanner.nextLine().split(" ");
		if (!restaurants.containsKey(data[0])) {
		    if (verbose) {
			System.out.println(String.format("Bad reservation at line %d", line));
		    }
		    return false;
		}
		r = restaurants.get(data[0]);
		Reservation res = new Reservation(data[0], data[1], Integer.parseInt(data[2]));
		if (!r.makeReservation(res)) {
		    unconfirmed++;
		}
	    }
	} catch (FileNotFoundException e) {
	    if (verbose) {
		System.out.println("File not found:" + inputFilename);
	    }
	    return false;
	} catch (NumberFormatException e) {
	    if (verbose) {
		System.out.println("Format error: line " + line);
	    }
	    return false;
	} catch (ArrayIndexOutOfBoundsException e) {
	    if (verbose) {
		System.out.println("Format error: line " + line);
	    }
	    return false;
	}


	File resultsFile = TestHelpers.getResultsFile(f);
	line = 0;
	try {
	    Scanner scanner = new Scanner(resultsFile);
	    int expected_unconfirmed = Integer.parseInt(scanner.nextLine());

	    if (unconfirmed != expected_unconfirmed) {
		if (verbose) {
		    System.out.println(String.format("Actual value for the number of unconfirmed reservations (%d) does not match the expected value of unconfirmed reservations (%d)", unconfirmed, expected_unconfirmed));
		}
		return false;
	    }

	    for (i = 0; i < numReservations - unconfirmed; i++) {
		line++;
		String[] values = scanner.nextLine().split(" ");
		String rname = values[0];
		String pname = values[1];
		if (!restaurants.containsKey(rname)) {
		    if (verbose) {
			System.out.println(String.format("Cannot find restaurant for %s", rname));
		    }
		    return false;
		}
		r = restaurants.get(rname);
		int[] tables = new int[values.length - 2];
		for (int j = 2; j < values.length; j++) {
		    tables[j - 2] = Integer.parseInt(values[j]);
		}

		if (!r.confirmReservation(pname, tables)) {
		    if (verbose) {
			System.out.println(String.format("Reservation for %s @ %s is not correct.", pname, rname));
		    }
		    return false;
		}
	    }
	    return true;
	} catch (FileNotFoundException e) {
	    if (verbose) {
		System.out.println("Results file not found:" + inputFilename);
	    }
	    return false;
	} catch (NumberFormatException e) {
	    if (verbose) {
		System.out.println("Results file format error: line " + line);
	    }
	    return false;
	} catch (ArrayIndexOutOfBoundsException e) {
	    if (verbose) {
		System.out.println("Results file format error: line " + line);
	    }
	    return false;
	}
    }


    public static void main(String[] args) {
	System.exit(TestHelpers.doTests(args, new Test5()));	
    }
}

File: cpp/src/Test5.cpp
#include "TestHelpers.h"
#include "Restaurant.h"
#include <cstdio>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <cctype>  
#include <unordered_map>

class Test {

static std::vector<std::string> getData(std::string& f) {
    std::string dataFileName = TestHelpers::getDataFile(f);
    std::vector<std::string> res;
    int num_values;

    std::ifstream ifs(dataFileName);

    if (ifs) {
        std::string communityName;
        ifs >> communityName >> num_values;
        res.push_back(communityName);
    } else {
        std::cout << dataFileName << " failed to be read" << std::endl;;
        return std::vector<std::string>();
    }

    for (int i = 0; i < num_values; ++i) {
        std::string temp;
        ifs >> temp;
        res.push_back(temp);
    }
    
    if (res.size() != num_values + 1) {
        std::cout << "Format error: Actual and expected number of values do not match: " << dataFileName << std::endl;;
        exit(1);
    }

    return res;
}  


public: static bool runTest(std::string f) {
    std::string dataFileName = TestHelpers::getDataFile(f);
    std::unordered_map<std::string, Restaurant*> Restaurants;
    int numOfRes = 0, numOfParties = 0;
    int unconfirmed = 0;
    int i;
    Restaurant* r;

    std::ifstream ifs(dataFileName);
    if (ifs) {
        ifs >> numOfRes >> numOfParties;
    } else {
        std::cout << dataFileName << " failed to be read" << std::endl;;
    }
    
    for (i = 0; i < numOfRes; ++i) {
        std::string resName;
        int numOfTables;
        ifs >> resName >> numOfTables;
        Restaurants[resName] = new Restaurant(resName, numOfTables);
    }


    for (i = 0; i < numOfParties; ++i) {
        std::string resName, partyName;
        int partySize;
        ifs >> resName >> partyName >> partySize;
        Reservation *Reserve = new Reservation(resName, partyName, partySize);
        if (Restaurants.find(resName) != Restaurants.end()) {
            r = Restaurants[resName];
            if (!r->makeReservation(*Reserve)) {
                unconfirmed++;
            }
        } else {
            std::cout << "Cannot find restaurant for " << resName << "when making reservation\n";
	    return false;
        }
    }

    std::string resultFileName = TestHelpers::getResultsFile(f);
    std::ifstream ifs2(resultFileName);
    int expectedUnconfirmed;
    if (ifs2) {
        ifs2 >> expectedUnconfirmed;
        ifs2.ignore();
        if (unconfirmed != expectedUnconfirmed) {
            std::cout << "Actual value for the number of unconfirmed reservations (" << unconfirmed << ") ";
            std::cout << "does not match the expected value of unconfirmed reservations (" << expectedUnconfirmed << ")\n";
	    return false;
        }

        for (size_t k = 0 ; k < numOfParties - expectedUnconfirmed; ++k) {
            std::string temp;
            std::getline(ifs2, temp);
            std::string resName, partyName;
            std::istringstream iss(temp);
            iss >> resName >> partyName;
            std::vector<int> tables;
            while (!iss.eof()) {
                int table;
                iss >> table;
                tables.push_back(table);
            }

            if (Restaurants.find(resName) != Restaurants.end()) {
                Restaurant* res = Restaurants[resName];
                if(!res->checkReservation(partyName, tables)) {
                    std::cout << "Reservation for " << partyName << " at " << resName << " is not correct.\n";
		    return false;
                }
            } else {
                std::cout << "Cannot find restaurant for " << resName << "\n";
		return false;
            }
        }
    } else {
        std::cout << resultFileName << " failed to be read" << std::endl;;
	return false;
    }
    return true;
}

};

int main (int argc, char* argv[]) {
    std::vector<std::string> files = TestHelpers::getFiles(argc, argv);
    if (files.size() == 0) {
        std::cout << "no test files found"; 
        exit(1);
    } 

    int count = 0;
    for (size_t i = 0; i < files.size(); ++i) {
        std::string f = files[i];
	std::cout << "Test " << f << ": ";	 	
        if (Test::runTest(f)) {
	  std::cout << "PASSED\n";	  
	  count++;
	} else {
	  std::cout << "FAILED\n";	  
	}
    }

    if (count == files.size()) {
      return 42;
    } else {
      return 0;
    }
}
